File: src/agent_deploy/adapt.py
#!/usr/bin/env python3
"""
Agent adaptation script - REFACTORED to use agent.json as primary source.

Converts a PilotDeck market agent into the target AI tool's native format.
Now prioritizes agent.json with instructions field, falls back to SKILL.md
for backward compatibility.

Usage:
    python adapt.py --agent-path <dir> --target <tool_name> [--output-dir <dir>]
    python adapt.py --agent-path <dir> --target all [--output-dir <dir>]
"""
import argparse
import json
import os
import re
import shutil
import sys
import yaml
from pathlib import Path
from typing import Any, Dict, Optional, Tuple


# ---------------------------------------------------------------------------
# Agent Descriptor - Unified internal format
# ---------------------------------------------------------------------------

class AgentDescriptor:
    """Unified agent representation from agent.json or SKILL.md."""

    def __init__(self, name: str, display_name: str, version: str,
                 description: str, instructions: str, capabilities: list = None,
                 compatibility: dict = None, metadata: dict = None):
        self.name = name
        self.display_name = display_name
        self.version = version
        self.description = description
        self.instructions = instructions
        self.capabilities = capabilities or []
        self.compatibility = compatibility or {}
        self.metadata = metadata or {}

    def to_dict(self) -> Dict:
        return {
            "name": self.name,
            "display_name": self.display_name,
            "version": self.version,
            "description": self.description,
            "instructions": self.instructions,
            "capabilities": self.capabilities,
            "compatibility": self.compatibility,
            "metadata": self.metadata,
        }


# ---------------------------------------------------------------------------
# Agent.json parser (NEW - Primary)
# ---------------------------------------------------------------------------

def load_agent_descriptor(agent_path: Path) -> AgentDescriptor:
    """Load agent from agent.json (primary) or SKILL.md (fallback).

    Priority:
    1. agent.json with instructions field
    2. agent.json + external instruction file
    3. SKILL.md (deprecated, backward compatibility)
    """
    agent_json_path = agent_path / "agent.json"
    skill_md_path = agent_path / "SKILL.md"

    # Try agent.json first
    if agent_json_path.exists():
        try:
            agent_json = json.loads(agent_json_path.read_text(encoding="utf-8"))
            return _parse_agent_json(agent_json, agent_path)
        except Exception as e:
            print(f"Warning: Failed to parse agent.json: {e}", file=sys.stderr)
            print(f"Falling back to SKILL.md...", file=sys.stderr)

    # Fallback to SKILL.md
    if skill_md_path.exists():
        print(f"[DEPRECATED] Using SKILL.md as fallback. "
              f"Consider migrating to agent.json with instructions field.",
              file=sys.stderr)
        return _parse_skill_md(skill_md_path)

    raise FileNotFoundError(
        f"No agent.json or SKILL.md found in {agent_path}. "
        f"Agent directory must contain at least one of these files."
    )


def _parse_agent_json(agent_json: Dict, agent_path: Path) -> AgentDescriptor:
    """Parse agent.json into AgentDescriptor."""

    # Support both new (identity) and old (flat) format
    identity = agent_json.get("identity", agent_json)

    name = identity.get("name", agent_path.name)
    display_name = identity.get("display_name", name)
    version = identity.get("version", "1.0.0")
    description = identity.get("description", "")

    # Extract instructions (CORE CHANGE)
    instructions = ""

    if "instructions" in agent_json:
        inst = agent_json["instructions"]

        if inst.get("source") == "inline":
            instructions = inst.get("content", "")

        elif inst.get("source") == "file":
            inst_file = inst.get("file", "")
            if inst_file:
                inst_path = agent_path / inst_file
                if inst_path.exists():
                    instructions = inst_path.read_text(encoding="utf-8")
                else:
                    print(f"Warning: instruction file not found: {inst_path}", file=sys.stderr)

    # Fallback to SKILL.md if no instructions in agent.json
    if not instructions:
        skill_md_path = agent_path / "SKILL.md"
        if skill_md_path.exists():
            print(f"[DEPRECATED] agent.json found but no instructions field. "
                  f"Falling back to SKILL.md.", file=sys.stderr)
            instructions = skill_md_path.read_text(encoding="utf-8")
            # Strip YAML frontmatter for cleaner output
            instructions = _strip_frontmatter(instructions)

    if not instructions:
        raise ValueError(
            f"No instructions found in agent.json or SKILL.md. "
            f"Add 'instructions' field to agent.json or create SKILL.md."
        )

    capabilities = agent_json.get("capabilities", [])
    compatibility = agent_json.get("compatibility", {})

    metadata = {
        "schema_version": agent_json.get("schema_version", "1.0"),
        "author": identity.get("author", ""),
        "license": identity.get("license", "MIT"),
    }

    return AgentDescriptor(
        name=name,
        display_name=display_name,
        version=version,
        description=description,
        instructions=instructions,
        capabilities=capabilities,
        compatibility=compatibility,
        metadata=metadata,
    )


def _strip_frontmatter(text: str) -> str:
    """Remove YAML frontmatter from markdown."""
    match = re.match(r'^---\s*\n(.*?)\n---\s*\n', text, re.DOTALL)
    if match:
        return text[match.end():].strip()
    return text


# ---------------------------------------------------------------------------
# SKILL.md parser (LEGACY - Backward compatibility)
# ---------------------------------------------------------------------------

def _parse_skill_md(filepath: Path) -> AgentDescriptor:
    """Parse a PilotDeck SKILL.md file (legacy format)."""
    if not filepath.exists():
        raise FileNotFoundError(f"SKILL.md not found at {filepath}")

    text = filepath.read_text(encoding="utf-8")

    frontmatter = {}
    body = text

    # Extract YAML frontmatter
    fm_match = re.match(r'^---\s*\n(.*?)\n---\s*\n', text, re.DOTALL)
    if fm_match:
        try:
            frontmatter = yaml.safe_load(fm_match.group(1)) or {}
        except yaml.YAMLError:
            frontmatter = {}
        body = text[fm_match.end():].strip()

    name = frontmatter.get("name", filepath.parent.name)
    description = frontmatter.get("description", "").strip()

    return AgentDescriptor(
        name=name,
        display_name=frontmatter.get("display_name", name),
        version=frontmatter.get("version", "1.0.0"),
        description=description,
        instructions=body,  # Full markdown body
        capabilities=[],
        compatibility={},
        metadata={"source": "skill_md", "frontmatter": frontmatter},
    )


# ---------------------------------------------------------------------------
# Adapters for each target tool - Now using AgentDescriptor
# ---------------------------------------------------------------------------

def adapt_for_codebuddy(descriptor: AgentDescriptor) -> Dict:
    """
    CodeBuddy format: .codebuddy/skills/<name>/SKILL.md
    YAML frontmatter + markdown body.
    """
    fm = {
        "name": descriptor.name,
        "description": descriptor.description or f"Auto-adapted agent: {descriptor.name}",
        "version": descriptor.version,
    }
    body = f"""# {descriptor.display_name}

{descriptor.instructions}

---
*Adapted from PilotDeck Market by agent-deploy v2.0*
"""
    return {
        "tool": "codebuddy",
        "format": "markdown_yaml_frontmatter",
        "target_dir": f".codebuddy/skills/{descriptor.name}/",
        "target_file": "SKILL.md",
        "content": _render_skill_md(fm, body),
    }


def adapt_for_codebuddy_agent(descriptor: AgentDescriptor) -> Dict:
    """
    CodeBuddy Agent format: .codebuddy/agents/<name>.md
    Plain markdown file (NEW - added per user request).
    """
    body = f"""# {descriptor.display_name}

**Version**: {descriptor.version}
**Description**: {descriptor.description}

{descriptor.instructions}

---
*Adapted from PilotDeck Market by agent-deploy v2.0*
"""
    return {
        "tool": "codebuddy_agent",
        "format": "markdown_file",
        "target_dir": ".codebuddy/agents/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_claude_code(descriptor: AgentDescriptor) -> Dict:
    """
    Claude Code format: .claude/commands/<name>.md
    Plain markdown file.
    """
    body = f"""# {descriptor.display_name}

**Description**: {descriptor.description}

{descriptor.instructions}

---
*Adapted from PilotDeck Market by agent-deploy v2.0*
"""
    return {
        "tool": "claude_code",
        "format": "markdown_file",
        "target_dir": ".claude/commands/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_cursor(descriptor: AgentDescriptor) -> Dict:
    """
    Cursor format: .cursor/commands/<name>.md
    Plain markdown file for custom slash commands.
    """
    body = f"""# {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "cursor",
        "format": "markdown_file",
        "target_dir": ".cursor/commands/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_copilot(descriptor: AgentDescriptor) -> Dict:
    """
    GitHub Copilot format: .github/agents/<name>.md
    Plain markdown file.
    """
    body = f"""# Agent: {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "github_copilot",
        "format": "markdown_file",
        "target_dir": ".github/agents/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_opencode(descriptor: AgentDescriptor) -> Dict:
    """
    OpenCode format: .opencode/commands/<name>.md
    Plain markdown file for custom commands.
    """
    body = f"""# {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "opencode",
        "format": "markdown_file",
        "target_dir": ".opencode/commands/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_windsurf(descriptor: AgentDescriptor) -> Dict:
    """
    Windsurf format: .windsurf/rules/<name>.md
    Plain markdown file.
    """
    body = f"""# {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "windsurf",
        "format": "markdown_file",
        "target_dir": ".windsurf/rules/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_trae(descriptor: AgentDescriptor) -> Dict:
    """
    Trae format: .trae/rules/<name>.md
    Plain markdown file.
    """
    body = f"""# {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "trae",
        "format": "markdown_file",
        "target_dir": ".trae/rules/",
        "target_file": f"{descriptor.name.replace(' ', '-')}.md",
        "content": body,
    }


def adapt_for_aider(descriptor: AgentDescriptor) -> Dict:
    """
    Aider format: CONVENTIONS.md (appended) or standalone .aider/ file.
    """
    body = f"""## Agent: {descriptor.display_name}

{descriptor.description}

{descriptor.instructions}
"""
    return {
        "tool": "aider",
        "format": "conventions_markdown",
        "target_dir": ".",
        "target_file": "CONVENTIONS.md",
        "content": body,
        "notes": "Append to existing CONVENTIONS.md or create new one",
    }


def adapt_for_agents_md(descriptor: AgentDescriptor) -> Dict:
    """
    Universal AGENTS.md format: append a section to AGENTS.md in project root.
    Supported by Cursor, Copilot, CodeBuddy, Claude Code, Windsurf, etc.
    """
    section = f"""## Agent: {descriptor.display_name}

**Description**: {descriptor.description}

{descriptor.instructions}

"""
    return {
        "tool": "agents_md",
        "format": "markdown_section",
        "target_dir": ".",
        "target_file": "AGENTS.md",
        "content": section,
        "notes": "Append to existing AGENTS.md or create new one. Compatible with all tools that support AGENTS.md.",
    }


# ---------------------------------------------------------------------------
# Adapter registry
# ---------------------------------------------------------------------------

ADAPTERS = {
    "codebuddy": adapt_for_codebuddy,
    "codebuddy_agent": adapt_for_codebuddy_agent,  # NEW
    "claude_code": adapt_for_claude_code,
    "cursor": adapt_for_cursor,
    "github_copilot": adapt_for_copilot,
    "opencode": adapt_for_opencode,
    "windsurf": adapt_for_windsurf,
    "trae": adapt_for_trae,
    "aider": adapt_for_aider,
    "agents_md": adapt_for_agents_md,
}


def _render_skill_md(frontmatter: Dict, body: str) -> str:
    """Render a SKILL.md file from frontmatter and body."""
    fm_yaml = yaml.dump(frontmatter, allow_unicode=True, default_flow_style=False).strip()
    return f"---\n{fm_yaml}\n---\n\n{body}"


def adapt_agent(agent_path: Path, target: str) -> Optional[Dict]:
    """Adapt a PilotDeck agent to a target tool's format.

    Now uses load_agent_descriptor() which prioritizes agent.json.
    """
    # Load unified AgentDescriptor
    descriptor = load_agent_descriptor(agent_path)

    if target == "all":
        results = {}
        for tool_name, adapter_fn in ADAPTERS.items():
            results[tool_name] = adapter_fn(descriptor)
        return {"multi": True, "results": results}

    adapter_fn = ADAPTERS.get(target)
    if not adapter_fn:
        print(f"Error: Unknown target tool '{target}'. Available: {', '.join(ADAPTERS.keys())}")
        return None

    return adapter_fn(descriptor)


def write_adapted(adapted: Dict, output_dir: Path):
    """Write adapted agent to the output directory."""
    if adapted.get("multi"):
        for tool_name, result in adapted["results"].items():
            tool_dir = output_dir / tool_name
            tool_dir.mkdir(parents=True, exist_ok=True)
            filepath = tool_dir / result["target_file"]
            filepath.parent.mkdir(parents=True, exist_ok=True)
            filepath.write_text(result["content"], encoding="utf-8")
            print(f"  -> {filepath}")
        return

    filepath = output_dir / adapted["target_file"]
    output_dir.mkdir(parents=True, exist_ok=True)
    filepath.write_text(adapted["content"], encoding="utf-8")
    print(f"  -> {filepath}")


def main():
    parser = argparse.ArgumentParser(description="Adapt PilotDeck agent to target AI tool format")
    parser.add_argument("--agent-path", required=True, help="Path to the agent directory (contains agent.json or SKILL.md)")
    parser.add_argument("--target", default="all",
                        help=f"Target tool: {', '.join(ADAPTERS.keys())}, or 'all'")
    parser.add_argument("--output-dir", default="auto-adapter-output",
                        help="Output directory for adapted agents")
    parser.add_argument("--format", choices=["json", "files"], default="files",
                        help="Output format: 'json' returns content as JSON, 'files' writes to disk")
    args = parser.parse_args()

    agent_path = Path(args.agent_path).resolve()
    if not agent_path.exists():
        print(f"Error: Agent path not found: {agent_path}")
        sys.exit(1)

    adapted = adapt_agent(agent_path, args.target)
    if adapted is None:
        sys.exit(1)

    if args.format == "json":
        print(json.dumps(adapted, indent=2, ensure_ascii=False))
    else:
        output_dir = Path(args.output_dir).resolve()
        write_adapted(adapted, output_dir)
        print(f"\nAdapted {args.target} to {output_dir}/")


if __name__ == "__main__":
    main()

File: node/src/adapt.ts
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { loadRegistry } from "./registry.js";
import * as yaml from "js-yaml";

/** Regex to detect YAML frontmatter delimited by --- */
const FRONTMATTER_RE = /^---\n([\s\S]*?)\n---\n/;

export interface AdaptationResult {
  /** The adapted markdown / yaml+markdown content. */
  content: string;
  /** Relative path where this file should be written. */
  target_file: string;
  /** Format hint: "markdown" or "yaml+markdown". */
  format: string;
  /** When true, content should be appended to an existing file. */
  append?: boolean;
  /** Agent name slug for path substitution. */
  slug?: string;
}

/**
 * Unified Agent Descriptor - internal representation.
 */
interface AgentDescriptor {
  name: string;
  displayName: string;
  version: string;
  description: string;
  instructions: string;
  capabilities: any[];
  compatibility: Record<string, any>;
  metadata: Record<string, any>;
}

/**
 * Load agent from agent.json (primary) or SKILL.md (fallback).
 *
 * Priority:
 * 1. agent.json with instructions field
 * 2. agent.json + external instruction file
 * 3. SKILL.md (deprecated, backward compatibility)
 */
function loadAgentDescriptor(agentPath: string): AgentDescriptor {
  const agentJsonPath = join(agentPath, "agent.json");
  const skillMdPath = join(agentPath, "SKILL.md");

  // Try agent.json first
  if (existsSync(agentJsonPath)) {
    try {
      const raw = readFileSync(agentJsonPath, "utf8");
      const agentJson = JSON.parse(raw);
      return parseAgentJson(agentJson, agentPath);
    } catch (err: any) {
      console.warn(`[WARN] Failed to parse agent.json: ${err.message}`);
      console.warn(`[WARN] Falling back to SKILL.md...`);
    }
  }

  // Fallback to SKILL.md
  if (existsSync(skillMdPath)) {
    console.warn(`[DEPRECATED] Using SKILL.md as fallback. Consider migrating to agent.json with instructions field.`);
    return parseSkillMd(skillMdPath);
  }

  throw new Error(
    `No agent.json or SKILL.md found in ${agentPath}. ` +
    `Agent directory must contain at least one of these files.`
  );
}

/**
 * Parse agent.json into AgentDescriptor.
 */
function parseAgentJson(agentJson: any, agentPath: string): AgentDescriptor {
  // Support both new (identity) and old (flat) format
  const identity = agentJson.identity || agentJson;

  const name = identity.name || agentPath.split("/").pop() || "agent";
  const displayName = identity.display_name || identity.displayName || name;
  const version = identity.version || "1.0.0";
  const description = identity.description || "";

  // Extract instructions (CORE CHANGE)
  let instructions = "";

  if (agentJson.instructions) {
    const inst = agentJson.instructions;

    if (inst.source === "inline") {
      instructions = inst.content || "";
    } else if (inst.source === "file") {
      const instFile = inst.file || "";
      if (instFile) {
        const instPath = join(agentPath, instFile);
        if (existsSync(instPath)) {
          instructions = readFileSync(instPath, "utf8");
        } else {
          console.warn(`[WARN] Instruction file not found: ${instPath}`);
        }
      }
    }
  }

  // Strategy 2: Generate from subagents (PilotDeck Agent format)
  if (!instructions && agentJson.subagents && agentJson.subagents.length > 0) {
    const entry = agentJson.entry?.main_subagent || agentJson.subagents[0].name;

    instructions = `# ${identity.display_name || identity.name}

${identity.description || ""}

## Workflows

This agent contains ${agentJson.subagents.length} sub-workflow(s):

${agentJson.subagents.map((sub: any) => `- **${sub.name}** (\`${sub.path}\`): ${sub.description || "No description"}`).join("\n")}

Entry workflow: **${entry}**

## Usage

This agent is based on PilotDeck workflow orchestration. See individual \`.yaml\` files for detailed configuration.
`;
    console.warn(`[INFO] Generated instructions from subagents for PilotDeck Agent format.`);
  }

  // Strategy 3: Fallback to SKILL.md
  if (!instructions) {
    const skillMdPath = join(agentPath, "SKILL.md");
    if (existsSync(skillMdPath)) {
      console.warn(`[DEPRECATED] agent.json found but no instructions field. Falling back to SKILL.md.`);
      instructions = readFileSync(skillMdPath, "utf8");
      // Strip YAML frontmatter for cleaner output
      instructions = stripFrontmatter(instructions);
    }
  }

  // Strategy 4: Fallback to README.md
  if (!instructions) {
    const readmePath = join(agentPath, "README.md");
    if (existsSync(readmePath)) {
      console.warn(`[FALLBACK] Using README.md as instructions source.`);
      instructions = readFileSync(readmePath, "utf8");
    }
  }

  if (!instructions) {
    throw new Error(
      `No instructions found. Tried:\n` +
      `1. agent.json instructions field\n` +
      `2. Generated from subagents\n` +
      `3. SKILL.md file\n` +
      `4. README.md file\n` +
      `Please add at least one of these to your agent.`
    );
  }

  const capabilities = agentJson.capabilities || [];
  const compatibility = agentJson.compatibility || {};

  const metadata = {
    schema_version: agentJson.schema_version || "1.0",
    author: identity.author || "",
    license: identity.license || "MIT",
  };

  return {
    name,
    displayName,
    version,
    description,
    instructions,
    capabilities,
    compatibility,
    metadata,
  };
}

/**
 * Strip YAML frontmatter from markdown.
 */
function stripFrontmatter(text: string): string {
  const match = text.match(FRONTMATTER_RE);
  if (match) {
    return text.slice(match[0].length).trim();
  }
  return text.trim();
}

/**
 * Parse SKILL.md (legacy format) into AgentDescriptor.
 */
function parseSkillMd(skillPath: string): AgentDescriptor {
  const raw = readFileSync(skillPath, "utf8");
  const { frontmatter, body } = parseFrontmatter(raw);

  const fm = frontmatter ? parseYamlFrontmatter(frontmatter) : {};
  const name = fm.name || skillPath.split("/").slice(-2, -1)[0] || "agent";
  const description = (fm.description || "").trim();

  return {
    name,
    displayName: fm.display_name || fm.displayName || name,
    version: fm.version || "1.0.0",
    description,
    instructions: body,
    capabilities: [],
    compatibility: {},
    metadata: { source: "skill_md", frontmatter: fm },
  };
}

/**
 * Extract YAML frontmatter (the block between --- delimiters) and the body.
 */
function parseFrontmatter(raw: string): { frontmatter: string | null; body: string } {
  const match = raw.match(FRONTMATTER_RE);
  if (match) {
    return {
      frontmatter: match[1].trim(),
      body: raw.slice(match[0].length).trim(),
    };
  }
  return { frontmatter: null, body: raw.trim() };
}

/**
 * Parse YAML frontmatter string into object (simplified).
 */
function parseYamlFrontmatter(yaml: string): Record<string, any> {
  const result: Record<string, any> = {};
  const lines = yaml.split("\n");
  for (const line of lines) {
    const match = line.match(/^(\w+):\s*(.*)$/);
    if (match) {
      const key = match[1];
      let value: any = match[2].trim();
      // Remove quotes
      if ((value.startsWith('"') && value.endsWith('"')) ||
          (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      result[key] = value;
    }
  }
  return result;
}

/**
 * Load and translate worker.yaml pipeline into executable CC prompt instructions.
 * Returns null if no worker.yaml found or parsing fails.
 */
function loadWorkerYamlAsPrompt(agentPath: string, subagentPath: string): string | null {
  const yamlPath = join(agentPath, subagentPath);
  if (!existsSync(yamlPath)) return null;

  let parsed: any;
  try {
    const content = readFileSync(yamlPath, "utf8");
    parsed = yaml.load(content);
  } catch {
    return null;
  }

  if (!parsed || !Array.isArray(parsed.pipeline)) return null;

  const toolInstructions: Record<string, (args: any) => string> = {
    write_file: (args: any) =>
      `Write the following content to file \`${args.path}\`:\n\`\`\`\n${args.content}\n\`\`\`${args.mode === "append" ? "\n(append to existing file)" : ""}`,
    web_fetch: (args: any) =>
      `Fetch URL via HTTP ${args.method || "GET"}: \`${args.url}\``,
    bash: (args: any) =>
      `Run shell command:\n\`\`\`bash\n${args.command}\n\`\`\``,
    read_file: (args: any) =>
      `Read file: \`${args.path}\``,
    glob: (args: any) =>
      `Find files matching pattern: \`${args.pattern}\``,
    llm_chat: (args: any) =>
      `Ask LLM: ${args.prompt}`,
    web_search: (args: any) =>
      `Search the web for: ${args.query}`,
  };

  // Extract all {{varName}} template variables from args, excluding steps.* and shared_context keys
  const varPattern = /\{\{([^}]+)\}\}/g;
  const allVars = new Set<string>();
  const sharedKeys = new Set(Object.keys(parsed.shared_context || {}));
  for (const step of parsed.pipeline) {
    const argsStr = JSON.stringify(step.args || {});
    let m: RegExpExecArray | null;
    while ((m = varPattern.exec(argsStr)) !== null) {
      const v = m[1].trim();
      if (!v.startsWith("steps.") && !v.startsWith("shared_context.") && !sharedKeys.has(v)) {
        allVars.add(v);
      }
    }
  }

  const paramSection = allVars.size > 0
    ? `## Parameters\n\nProvide the following values when invoking this agent (use \`$ARGUMENTS\` or pass as key=value):\n\n${[...allVars].map(v => `- \`${v}\``).join("\n")}\n\n`
    : "";

  const steps: string[] = parsed.pipeline.map((step: any, i: number) => {
    const num = i + 1;
    const toolFn = (toolInstructions as any)[step.tool];
    const instruction = toolFn
      ? toolFn(step.args || {})
      : `Call tool \`${step.tool}\` with args: ${JSON.stringify(step.args || {})}`;
    const outputNote = step.output ? ` Save result as \`${step.output}\`.` : "";
    return `**Step ${num}: ${step.step}**\n${instruction}${outputNote}`;
  });

  const sharedCtx = parsed.shared_context
    ? `\n\n**Shared context**: ${JSON.stringify(parsed.shared_context)}`
    : "";

  return paramSection + steps.join("\n\n") + sharedCtx;
}

/**
 * Generate a URL-safe slug from an agent name.
 */
function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

/**
 * Read agent from directory and adapt it for the target tool.
 * Now uses loadAgentDescriptor() which prioritizes agent.json.
 *
 * @param agentPath - Path to the agent directory (must contain agent.json or SKILL.md).
 * @param target    - Registry key of the target tool (e.g. "opencode").
 * @returns AdaptationResult with the transformed content and target path.
 */
export function adaptAgent(agentPath: string, target: string, targetFile?: string): AdaptationResult {
  // Load unified AgentDescriptor
  const descriptor = loadAgentDescriptor(agentPath);
  const slug = slugify(descriptor.name);

  const registry = loadRegistry();
  const toolConfig = registry.tools[target];
  if (!toolConfig) {
    return {
      content: `# Error\n\nUnknown target tool: ${target}`,
      target_file: `unknown/${target}.md`,
      format: "markdown",
      slug,
    };
  }

  let result: AdaptationResult;

  // Adapt based on target tool
  switch (target) {
    case "opencode":
    case "cursor": {
      // Strip YAML frontmatter, wrap as a markdown command.
      const header = `# ${descriptor.displayName}\n\n`;
      const content = header + descriptor.instructions;
      result = {
        content,
        target_file: `.${target}/commands/${slug}.md`,
        format: "markdown",
        slug,
      };
      break;
    }

    case "codebuddy": {
      // Keep YAML frontmatter, add codebuddy-specific header.
      let out = `---\nname: ${descriptor.name}\nversion: ${descriptor.version}\ndescription: ${descriptor.description}\n---\n\n`;
      out += `<!-- codebuddy:skill name="${descriptor.name}" -->\n\n`;
      out += descriptor.instructions;
      result = {
        content: out,
        target_file: `.codebuddy/skills/${slug}/SKILL.md`,
        format: "yaml+markdown",
        slug,
      };
      break;
    }

    case "codebuddy_agent": {
      // CodeBuddy Agent format - plain markdown with optional pipeline info
      const agentJsonPath = join(agentPath, "agent.json");
      let pipelineInfo = "";

      // Check for worker.yaml pipeline (direct or via subagents)
      if (existsSync(agentJsonPath)) {
        try {
          const agentJson = JSON.parse(readFileSync(agentJsonPath, "utf8"));
          let pipelinePath: string | null = null;

          // Check direct worker.yaml in agent dir
          const directWorkerPath = join(agentPath, "worker.yaml");
          if (existsSync(directWorkerPath)) {
            pipelinePath = "worker.yaml";
          }
          // Check subagents for entry point
          else if (agentJson.subagents && agentJson.subagents.length > 0) {
            const entryName = agentJson.entry?.main_subagent || agentJson.subagents[0].name;
            const entrySubagent = agentJson.subagents.find((s: any) => s.name === entryName) || agentJson.subagents[0];
            if (entrySubagent && entrySubagent.path) {
              pipelinePath = entrySubagent.path;
            }
          }

          if (pipelinePath) {
            const pipelinePrompt = loadWorkerYamlAsPrompt(agentPath, pipelinePath);
            if (pipelinePrompt) {
              pipelineInfo = `## Pipeline\n\n${pipelinePrompt}\n\n`;
            }
          }
        } catch {
          // ignore, fall through to basic instructions
        }
      }

      const content =
        `# ${descriptor.displayName}\n\n` +
        `**Version**: ${descriptor.version}\n` +
        `**Description**: ${descriptor.description}\n\n` +
        pipelineInfo +
        `${descriptor.instructions}\n\n` +
        `---\n` +
        `*Adapted from Agent Market by agent-deploy v${descriptor.version}*\n`;
      result = {
        content,
        target_file: `.codebuddy/agents/${slug}.md`,
        format: "markdown",
        slug,
      };
      break;
    }

    case "claude_code": {
      // Check if agent has subagents with worker.yaml pipeline — translate to executable prompt
      const agentJsonPath = join(agentPath, "agent.json");
      let pipelinePrompt: string | null = null;
      if (existsSync(agentJsonPath)) {
        try {
          const agentJson = JSON.parse(readFileSync(agentJsonPath, "utf8"));
          if (agentJson.subagents && agentJson.subagents.length > 0) {
            const entrySubagentName = agentJson.entry?.main_subagent || agentJson.subagents[0].name;
            const entrySubagent = agentJson.subagents.find((s: any) => s.name === entrySubagentName) || agentJson.subagents[0];
            pipelinePrompt = loadWorkerYamlAsPrompt(agentPath, entrySubagent.path);
          }
        } catch {
          // fall through to descriptor.instructions
        }
      }

      const body = pipelinePrompt
        ? `## Description\n\n${descriptor.description}\n\n## Steps\n\nExecute the following steps in order:\n\n${pipelinePrompt}`
        : `## Description\n\n${descriptor.description}\n\n${descriptor.instructions}`;

      const ccContent = `# /${slug} — ${descriptor.displayName}\n\n${body}`;
      result = {
        content: ccContent,
        target_file: `.claude/commands/${slug}.md`,
        format: "markdown",
        slug,
      };
      break;
    }

    case "github_copilot": {
      // Markdown with an agent-style prompt header.
      const ghContent = `# ${descriptor.displayName}\n\n> AI agent prompt – generated by agent-deploy v2.0\n\n${descriptor.description}\n\n${descriptor.instructions}`;
      result = {
        content: ghContent,
        target_file: `.github/agents/${slug}.md`,
        format: "markdown",
        slug,
      };
      break;
    }

    case "windsurf":
    case "trae": {
      // Rules-style format — add a small header block.
      const rulesContent = `# Rule: ${descriptor.displayName}\n\n${descriptor.description}\n\n${descriptor.instructions}`;
      const dir = `.${target}/rules`;
      result = {
        content: rulesContent,
        target_file: `${dir}/${slug}.md`,
        format: "markdown",
        slug,
      };
      break;
    }

    case "aider": {
      // Append-mode: CONVENTIONS.md section.
      const aiderSection = `\n\n---\n## ${descriptor.displayName}\n\n${descriptor.description}\n\n${descriptor.instructions}`;
      result = {
        content: aiderSection,
        target_file: "CONVENTIONS.md",
        format: "markdown",
        append: true,
        slug,
      };
      break;
    }

    case "agents_md": {
      // Append-mode: AGENTS.md section.
      const agentsSection = `\n\n---\n## ${descriptor.displayName}\n\n**Description**: ${descriptor.description}\n\n${descriptor.instructions}`;
      result = {
        content: agentsSection,
        target_file: "AGENTS.md",
        format: "markdown",
        append: true,
        slug,
      };
      break;
    }

    default:
      throw new Error(`No adaptation strategy for tool: ${target}`);
  }

  // Override target_file if explicitly provided
  if (targetFile) {
    result.target_file = targetFile;
  }

  return result;
}
